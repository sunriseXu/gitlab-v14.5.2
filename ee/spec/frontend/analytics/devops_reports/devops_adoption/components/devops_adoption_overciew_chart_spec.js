import { createLocalVue, shallowMount } from '@vue/test-utils';
import VueApollo from 'vue-apollo';
import { GlStackedColumnChart } from '@gitlab/ui/dist/charts';
import ChartSkeletonLoader from '~/vue_shared/components/resizable_chart/skeleton_loader.vue';
import DevopsAdoptionOverviewChart from 'ee/analytics/devops_reports/devops_adoption/components/devops_adoption_overview_chart.vue';
import getSnapshotsQuery from 'ee/analytics/devops_reports/devops_adoption/graphql/queries/devops_adoption_overview_chart.query.graphql';
import createMockApollo from 'helpers/mock_apollo_helper';
import { namespaceWithSnapotsData } from '../mock_data';

const localVue = createLocalVue();
localVue.use(VueApollo);

const mockWithData = jest.fn().mockResolvedValue(namespaceWithSnapotsData);

describe('DevopsAdoptionOverviewChart', () => {
  let wrapper;

  afterEach(() => {
    wrapper.destroy();
  });

  const createComponent = ({ stubs = {}, mockSnapshotsQuery = mockWithData, data = {} } = {}) => {
    const handlers = [[getSnapshotsQuery, mockSnapshotsQuery]];

    wrapper = shallowMount(DevopsAdoptionOverviewChart, {
      localVue,
      provide: {
        groupGid:
          namespaceWithSnapotsData.data.devopsAdoptionEnabledNamespaces.nodes[0].namespace.id,
      },
      apolloProvider: createMockApollo(handlers),
      data() {
        return {
          ...data,
        };
      },
      stubs,
    });
  };

  describe('default state', () => {
    beforeEach(() => {
      createComponent();
    });

    it('does not display the chart loader', () => {
      expect(wrapper.findComponent(ChartSkeletonLoader).exists()).toBe(false);
    });

    it('displays the chart', () => {
      expect(wrapper.findComponent(GlStackedColumnChart).exists()).toBe(true);
    });

    it('computes the correct series data', () => {
      expect(wrapper.findComponent(GlStackedColumnChart).props('bars')).toMatchSnapshot();
    });
  });

  describe('loading', () => {
    it('displays the chart loader', () => {
      createComponent({});
      expect(wrapper.findComponent(ChartSkeletonLoader).exists()).toBe(true);
    });

    it('does not display the chart', () => {
      createComponent({});
      expect(wrapper.findComponent(GlStackedColumnChart).exists()).toBe(false);
    });
  });

  describe('chart tooltip', () => {
    beforeEach(() => {
      const mockParams = {
        value: 'Jan',
        seriesData: [{ dataIndex: 0 }],
      };

      createComponent({
        stubs: {
          GlStackedColumnChart: {
            props: ['formatTooltipText'],
            mounted() {
              this.formatTooltipText(mockParams);
            },
            template: `
                <div>
                  <slot name="tooltip-title"></slot>
                  <slot name="tooltip-content"></slot>
                </div>`,
          },
        },
        data: {
          devopsAdoptionEnabledNamespaces: {
            nodes: namespaceWithSnapotsData.data.devopsAdoptionEnabledNamespaces.nodes,
          },
        },
      });
    });

    it('displays the tooltip information correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
