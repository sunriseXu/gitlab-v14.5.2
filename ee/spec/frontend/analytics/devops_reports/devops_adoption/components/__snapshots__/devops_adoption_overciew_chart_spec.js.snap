// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevopsAdoptionOverviewChart chart tooltip displays the tooltip information correctly 1`] = `
<div>
  <h4>
    Adoption over time
  </h4>
   
  <div
    bars="[object Object],[object Object],[object Object]"
    custom-palette="#b24800,#0094b6,#487900"
    group-by="January"
    option="[object Object]"
    presentation="tiled"
    x-axis-title=""
    x-axis-type="category"
    y-axis-title="Total number of features adopted"
  >
    
      Jan
     
    <div
      class="gl-mb-3"
    >
      <div
        class="gl-font-weight-bold"
      >
        Dev
      </div>
       
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          variant="warning"
        />
        
              Approvals
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          variant="warning"
        />
        
              Code owners
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          variant="warning"
        />
        
              Issues
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="warning"
        />
        
              MRs
            
      </div>
    </div>
    <div
      class="gl-mb-3"
    >
      <div
        class="gl-font-weight-bold"
      >
        Sec
      </div>
       
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="info"
        />
        
              DAST
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="info"
        />
        
              Dependency Scanning
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="info"
        />
        
              Fuzz Testing
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="info"
        />
        
              SAST
            
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="gl-font-weight-bold"
      >
        Ops
      </div>
       
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="success"
        />
        
              Deploys
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          variant="success"
        />
        
              Pipelines
            
      </div>
      <div>
        <devops-adoption-table-cell-flag-stub
          class="gl-mr-3"
          enabled="true"
          variant="success"
        />
        
              Runners
            
      </div>
    </div>
  </div>
</div>
`;

exports[`DevopsAdoptionOverviewChart default state computes the correct series data 1`] = `
Array [
  Object {
    "data": Array [
      1,
    ],
    "name": "Dev",
  },
  Object {
    "data": Array [
      4,
    ],
    "name": "Sec",
  },
  Object {
    "data": Array [
      2,
    ],
    "name": "Ops",
  },
]
`;
